app/kuma-dp/pkg/dataplane/dnsproxy/component.go

```diff
@@ -0,0 +1,192 @@
+package dnsproxy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/miekg/dns"
+
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/pkg/dns/dpapi"
+)
+
+var log = core.Log.WithName("embeededdns")
+
+type Server struct {
+	address          string
+	port             uint32
+	done             chan struct{}
+	dnsMap           atomic.Pointer[dnsMap]
+	metrics          *metrics
+	upstreamHostPort string
+}
+
+func NewServer(address string, port uint32) *Server {
```

Comment:
do we have a test for this? just starting the server, seeding it and having a client query it and seeing if the results match?

Reply:
done